lib/navigation/preprocess_imu.cpp

```diff
@@ -1,56 +1,137 @@
-ImuPreprocessor::ImuPreprocessor()
+ImuPreprocessor::ImuPreprocessor(core::ILogger &logger)
+    : log_(logger),
+      num_outliers_per_imu_({0, 0, 0, 0}),
+      are_imus_reliable_({true, true, true, true}),
+      num_reliable_accelerometers_(core::kNumImus)
 {
-  // TODOLater: implement
 }
 
-core::ImuData ImuPreprocessor::processData(const core::RawImuData raw_imu_data)
+std::optional<core::ImuData> ImuPreprocessor::processData(const core::RawImuData raw_imu_data)
 {
-  /*
-  TODOLater: implement
-  basic plan:
-  - call detectOutliers. Return of that function is return of this function
-  - call checkReliable
-
-  */
-  return {0, 0, 0, 0};
+  core::ImuData imu_data;
+  // TODOLater : check on direction of travel
+  core::Float magnitude;
+  for (std::size_t i = 0; i < core::kNumImus; ++i) {
+    magnitude = 0;
+    for (std::size_t j = 0; j < 3; ++j) {
+      magnitude += std::pow(raw_imu_data.at(i).at(j), 2);
+    }
+    imu_data.at(i) = std::sqrt(magnitude);
+  }
+
+  const core::ImuData accelerometer_data = detectOutliers(imu_data);
+  checkReliable();
+  return accelerometer_data;
+}
+
+core::ImuData ImuPreprocessor::detectOutliers(const core::ImuData imu_data)
+{
+  core::ImuData accelerometer_data;
+  for (std::size_t i; i < core::kNumImus; ++i) {
+    accelerometer_data.at(i) = imu_data.at(i);
+  }
+
+  Quartiles quartiles;
+  if (num_reliable_accelerometers_ == core::kNumImus) {
+    quartiles = getQuartiles(imu_data);
+  } else {
+    std::array<core::Float, core::kNumImus - 1> faulty_data;
+    std::size_t index;
+    for (std::size_t i = 0; i < core::kNumImus - 1; ++i) {
+      if (!are_imus_reliable_.at(i)) { index = i; }
+    }
+    for (std::size_t i = 0; i < core::kNumImus - 1; ++i) {
+      if (i >= index) {
+        faulty_data.at(i) = imu_data.at(i + 1);
+      } else {
+        faulty_data.at(i) = imu_data.at(i);
+      }
+    }
+    quartiles = getQuartiles(faulty_data);
+  }
+
+  const core::Float iqr = quartiles.q3 - quartiles.q1;
+  core::Float lower_bound;
+  core::Float upper_bound;
+
+  // TODOLater : Check these values
+  if (num_reliable_accelerometers_ == core::kNumImus) {
+    lower_bound = quartiles.median - 1.5 * iqr;
+    upper_bound = quartiles.median + 1.5 * iqr;
+  } else {
+    lower_bound = quartiles.median - 1.2 * iqr;
+    upper_bound = quartiles.median + 1.2 * iqr;
+  }
+
+  for (size_t i = 0; i < core::kNumImus; ++i) {
+    // converts outliers or unreliables to medians, updates number of consecutive outliers for each
+    // sensor
+    if (are_imus_reliable_.at(i) == false) {
+      accelerometer_data.at(i) = quartiles.median;
+    } else if (imu_data.at(i) < lower_bound || imu_data.at(i) > upper_bound) {
+      accelerometer_data.at(i) = quartiles.median;
+      ++num_outliers_per_imu_.at(i);
+    } else {
+      num_outliers_per_imu_.at(i) = 0;
+    }
+  }
+  return accelerometer_data;
 }
 
-core::ImuData ImuPreprocessor::detectOutliers(const core::RawImuData imu_data)
+template<std::size_t N>
```

Comment:
My understanding is that the templated functions should only be in header files. @mifrandir correct me if I'm wrong, similar to how the mux was done.

Reply:
You are right!

Reply:
Why do we want this template in the first place?

Also, this implementation is far from being asymptotically optimal so I'd like to see a comment noting that and what runtime we currently have. I'm pretty sure this can be done in linear time.

Reply:
In this implementation, N is either kNumImus or kNumImus-1. Currently this is either 3 or 4 so should run fairly quickly? The template allows the number of imus to be variable so currently the options seem to be either repetetive code or  template function. Is there a nicer way of doing this? Given that N is either 3 or 4, does this not still run in relatively constant time?